### PR TITLE
Add Restart task to CassandraTasks

### LINKED ISSUE #385: K8SSAND-1707 ⁃ Add task for rolling restart
**What is missing?**
At the moment, to do a rolling restart, we must modify the CassandraDatacenter spec. We want a task that would do the same without requiring the modification of Spec.

**Why do we need it?**
As noted in issue https://github.com/k8ssandra/cass-operator/issues/102 we should not be modifying the spec for operations. This is one of operations that should be migrated outside.

**Environment**

* Cass Operator version:

    `Insert image tag or Git SHA here`
    
    **Anything else we need to know?**:

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-1707) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-1707
┆priority: Medium

### PR DESCRIPTION
**Pull request by burmanm**
**What this PR does**:
Creates a new option that does rolling restart outside cads-operator's controller. 

**Which issue(s) this PR fixes**:
Fixes #385

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CHANGELOG.md updated (not required for documentation PRs)
- [x] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

**Comment by bradfordcp** [comment]
I see there is a `keyspaceArgument` in the changes, does this relate to the restart?

File: apis/control/v1alpha1/cassandratask_types.go [hunk 1]
```diff
@@ -69,6 +73,12 @@ const (
 	CommandScrub           CassandraCommand = "scrub"
 )
 
+const (
+	KeyspaceArgument         string = "keyspace_name"
```
Comment (bradfordcp) [inline apis/control/v1alpha1/cassandratask_types.go:77]: How is this related to a restart?

File: controllers/control/jobs.go [hunk 2]
```diff
@@ -3,22 +3,27 @@ package control
...
+	api "github.com/k8ssandra/cass-operator/apis/control/v1alpha1"
 )
 
 // Cleanup functionality
 
 func callCleanup(nodeMgmtClient httphelper.NodeMgmtClient, pod *corev1.Pod, taskConfig *TaskConfiguration) (string, error) {
 	// TODO Add more arguments configurations
 	keyspaceName := ""
-	if keyspace, found := taskConfig.Arguments["keyspace_name"]; found {
+	if keyspace, found := taskConfig.Arguments[api.KeyspaceArgument]; found {
```
Comment (bradfordcp) [inline controllers/control/jobs.go:26]: How is this related to a restart?

File: controllers/control/jobs.go [hunk 3]
```diff
@@ -27,7 +32,7 @@ func callCleanup(nodeMgmtClient httphelper.NodeMgmtClient, pod *corev1.Pod, task
 func callCleanupSync(nodeMgmtClient httphelper.NodeMgmtClient, pod *corev1.Pod, taskConfig *TaskConfiguration) error {
 	// TODO Add more arguments configurations
 	keyspaceName := ""
-	if keyspace, found := taskConfig.Arguments["keyspace_name"]; found {
+	if keyspace, found := taskConfig.Arguments[api.KeyspaceArgument]; found {
```
Comment (bradfordcp) [inline controllers/control/jobs.go:35]: How is this related to a restart?

File: controllers/control/jobs.go (hunk 2 above)
Comment (burmanm) [reply controllers/control/jobs.go:26]: This is mostly superceded by #395 , just to ensure we don't have magic strings around the code. Same for previous comment.

File: controllers/control/jobs.go (hunk 3 above)
Comment (burmanm) [reply controllers/control/jobs.go:35]: And same here

File: controllers/control/cassandratask_controller.go [hunk 4]
```diff
@@ -273,7 +275,17 @@ func (r *CassandraTaskReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
 			cleanup(taskConfig)
 		case api.CommandRestart:
-			r.restart(taskConfig)
+			// This job is targeting StatefulSets and not Pods
+			sts, err := r.getDatacenterStatefulSets(ctx, dc)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			res, err = r.restartSts(ctx, sts, taskConfig)
```
Comment (adejanovski) [inline controllers/control/cassandratask_controller.go:284]: The `res` value doesn't seem to be used at all. Maybe that's what could be causing prematurely setting the completion time?

File: controllers/control/cassandratask_controller_test.go [hunk 5]
```diff
@@ -158,254 +223,361 @@ func waitForTaskCompletion(taskKey types.NamespacedName) *api.CassandraTask {
...
+				Expect(k8sClient.Create(context.Background(), task)).Should(Succeed())
 
-			// Recreate it so the process "finishes"
-			createPod(testNamespaceName, fmt.Sprintf("test-%s", testDatacenterName), testDatacenterName, 2)
+				Eventually(func() bool {
+					Expect(k8sClient.Get(context.TODO(), stsKey, &sts)).Should(Succeed())
+					_, found := sts.Spec.Template.ObjectMeta.Annotations[api.RestartedAtAnnotation]
+					return found
+				}).Should(BeTrue())
+
```
Comment (adejanovski) [inline controllers/control/cassandratask_controller_test.go:508]: You could maybe add a check here to see if the completion time is still `nil`.

**Review by adejanovski — CHANGES_REQUESTED** [review]
The code mostly looks good to me, aside from a Result return that isn't used when invoking the rolling restart on the STS, unless I'm missing something.

It could be causing the only thing that looks problematic when testing this feature, which is that the completion time gets set right away when the CassandraTask gets created.
We'd need the completion time to be set only when the rolling restart is completed.

File: controllers/control/cassandratask_controller.go (hunk 4 above)
Comment (burmanm) [reply controllers/control/cassandratask_controller.go:284]: Isn't that res used on the line 333? I don't see it being overwritten anywhere.

File: controllers/control/cassandratask_controller_test.go (hunk 5 above)
Comment (burmanm) [reply controllers/control/cassandratask_controller_test.go:508]: Done.

File: controllers/control/cassandratask_controller.go (hunk 4 above)
Comment (adejanovski) [reply controllers/control/cassandratask_controller.go:284]: Isn't it overwritten on line 320?

```
res, failed, completed, err = r.reconcileEveryPodTask(ctx, dc, taskConfig)
```

Comment (adejanovski) [reply controllers/control/cassandratask_controller.go:284]: Or is `break JobDefinition` sending us straight to line 333?

Comment (burmanm) [reply controllers/control/cassandratask_controller.go:284]: break JobDefinition should break us from the for() clause and thus line 320 should never be run.

**Review by adejanovski — APPROVED** [review]
Looks good now 👍
